### DISCUSSION TITLE
How to calibrate wflow results?

### DISCUSSION THREAD

**Discussion by arsalansaghir1** [Deltares/Wflow.jl#292] [Q&A] [answered]
I just ran a model in wflow (in Julia) for my area of interest. but the results are too far off. I read about the calibration of results through wflow but there was no clear explanation. Could you guide me in detail on how to calibrate the results?
Are there any scripts available for calibration?

**Comment by verseve** [comment, marked as answer]
There are not (yet) general calibration scripts available.

You could first check the influence of the `KsatHorFrac` parameter, default value from hydromt_wflow is 100.  For steep catchments this value is often too large, resulting in underestimation of peak discharge values and overestimation of baseflow. If peaks are too high and baseflow is too low, increasing the `KsatHorFrac` value is most probably required. Wflow_sbm is also quite sensitive to the `f` parameter, that controls the exponential decline of the vertical hydraulic conductivity at the soil surface (`kv₀`). Note that `KsatHorFrac` is applied to the model parameter `kv₀`.  Also some users calibrate the saturated water content `θₛ`.

Reservoir parameters from hydromt_wflow should be considered a first rough estimate, this is especially valid for the  `demand` and `maxrelease` model parameters.

For some info on sensitive model parameters you could check the first reference in the list below. Of course you could also run your own model parameter sensitivity analysis. Other model parameters to consider are parameters related to snow (e.g. `TT`), transpiration ( e.g. `rootingdepth`), leakage to deep groundwater (`maxleakage`) and routing (e.g. manning's roughness).  

A couple of references that may help:
https://agupubs.onlinelibrary.wiley.com/doi/full/10.1029/2019WR026807
https://gmd.copernicus.org/preprints/gmd-2022-182/
https://hess.copernicus.org/articles/26/4407/2022/hess-26-4407-2022.pdf
https://www.sciencedirect.com/science/article/pii/S2214581821000239?via%3Dihub

**Comment by arsalansaghir1** [comment]
In my case, peak value is too high like 1500 cumecs and baseflow values are too low ranging from 0-5 cumecs. What value of KsatHorFrac should I take in order to calibrate these kind of results?

**Comment by JoostBuitink** [comment]
As all catchments have different characteristics and have different responses, it is difficult to provide a value (hence why we use a default value of 100). The best way to find the best value - and how calibration is typically performed - is by running several simulations with different values for the parameters you want to calibrate. Based on performance metrics such as the Kling-Gupta efficiency, you can select which parameter value provides the most accurate discharge simulations. 

You can easily change the model parameter values in the toml file, see the examples in the docs: https://deltares.github.io/Wflow.jl/dev/user_guide/step2_settings_file/#Modify-parameters

**Comment by arsalansaghir1** [comment]
I changed the **### KsatHorFrac** value from default 100 to 50 but it caused very little change in the discharges like average 5 cumecs change. It is very less. Do I need to change other parameters as well in order to observe a noticeable change?

**Comment by verseve** [comment]
You mentioned that simulated peak discharge is too high, and simulated baseflow is too low. In that case you need to increase the `KsatHorFrac` value, I would recommend to start with 250. Also, I would recommend to check the `KsatVer` values in your model parameter NetCDF file (staticmaps.nc), perhaps these values are already quite low?

**Comment by arsalansaghir1** [comment]
These are the KsatVer values in my staticmaps nc file.
![Screenshot 2023-03-22 134325](https://user-images.githubusercontent.com/127018672/226847874-0afae7f9-1c91-4c99-9d5d-85b1d4a61561.jpg)

**Comment by verseve** [comment]
In the TOML file, there is a mapping between the internal `kv₀` model parameter to the external NetCDF parameter, like:

```
[input.vertical] 
"kv₀" = "KsatVer"
```

So, you could inspect the staticmaps `KsatVer` layer. `KsatHorFrac` is applied to `kv₀` for the **horizontal** hydraulic conductivity. The `KsatVer` values may give an indication of an appropriate `KsatHorFrac` value.

**Comment by arsalansaghir1** [comment]
Okay Thank you I'll look into this.
Also can you tell me if the data from era5 needs to be processed beforehand? I'm running the simulation for 5 years and I'm using hourly data from era5 but while preparing files in hydromt, it gives allocation error (ram exceeds). Whats the solution for this?
![Screenshot 2023-03-22 163112](https://user-images.githubusercontent.com/127018672/226891441-d746d8d7-4280-4435-80c4-4b4c42272053.jpg)

**Comment by verseve** [comment]
>  Also can you tell me if the data from era5 needs to be processed beforehand? I'm running the simulation for 5 years and I'm using hourly data from era5 but while preparing files in hydromt, it gives allocation error (ram exceeds). Whats the solution for this?

Yes, era5 data needs to be processed beforehand, because the Wflow model grid resolution is different from era5.

Could you please report the issue (ERROR) in https://github.com/Deltares/hydromt_wflow/issues (because it is specifically about hydromt and not Wflow.jl)?

**Comment by JoostBuitink** [comment]
> I'm running the simulation for 5 years and I'm using hourly data from era5 but while preparing files in hydromt, it gives allocation error (ram exceeds). Whats the solution for this? 

(This is indeed a hydromt_wflow issue and not a Wflow.jl issue, but replying here anyway as this is a known problem but easily solved by splitting the produced forcing files)

If you run into memory errors, you can fix this by telling hydromt_wflow to write the forcing files per month (for example). The `wflow_update_forcing.ini` file in hydromt_wflow/examples shows you how to do this. See these lines: https://github.com/Deltares/hydromt_wflow/blob/main/examples/wflow_update_forcing.ini#L19-L20

**Comment by arsalansaghir1** [comment]
I changed the constant parameter to extremes. Like KsatHorFrac to 10000. But I'm seeing little to no change in my output flows. Why my output flow is not being affected? and how can I reduce the peaks in order to calibrate. My parameters values are as follows:

[setup_constant_pars]
KsatHorFrac = 10000
Cfmax = 0.6
cf_soil = 0.038
EoverR = 0.05
InfiltCapPath = 5
InfiltCapSoil = 2500
MaxLeakage = 0
rootdistpar = -500
TT = 0
TTI = 2
TTM = 0
WHC = 0.1
G_Cfmax = 5.3
G_SIfrac = 0.002
G_TT = 1.3

**Comment by verseve** [comment]
And you did update your wflow model with these settings (with hydromt_wflow)?

It is probably easier to apply these uniform changes in the Wflow settings file (TOML file):

```
[input.lateral.subsurface]
ksathorfrac.value = 1000
```

Generally changing the `ksathorfrac` parameter has an affect on the output flows. What is happening when you set the value much lower, for example at 1.0?

I think in you model setup there was also an issue with leaf area index climatology. Has this been resolved now? If this is not included correctly, interception could be too low, which could also result in high peaks.

As mentioned before you could also look into other parameter values, for example the `f` parameter that controls the exponential decline of vertical hydraulic conductivity. Soil thickness may also play a role, for example shallow soil depth (less storage) can also affect peak discharges. And if snow is an important process in the catchment, you could also look into snow parameters.

**Comment by arsalansaghir1** [comment]
When I give values directly in the toml file, it gives parse error in Julia that parameter already has a value. How to resolve this?

![Screenshot 2023-04-03 115951](https://user-images.githubusercontent.com/127018672/229435060-4055e434-d847-4287-b87c-b7ea862c6ad2.jpg)

**Comment by verseve** [comment]
You need to comment out the mapping to the external netCDF variable:
```
[input.lateral.subsurface]
#ksathorfrac = "KsatHorFrac"
ksathorfrac.value = 0.1
```

**Comment by abdulrehman565** [comment]
How KsatVer values are calculated by default when staticmaps.nc files is prepared? I want to understand how this value is calculated for multiple depths

**Comment by verseve** [comment]
These are calculated with the Brakensiek PTF, see also the hydromt_wflow [docs](https://deltares.github.io/hydromt_wflow/latest/_generated/hydromt_wflow.WflowModel.setup_soilmaps.html#hydromt_wflow.WflowModel.setup_soilmaps) and this [paper](https://agupubs.onlinelibrary.wiley.com/doi/full/10.1029/2019WR026807) for more detailed info.

**Comment by Xervou** [comment]
I would like to ask the author if he can improve this problem in the future?
I'm currently experiencing unusually high mainstream runoff.
But after modifying the KsatHorFrac value to 10000, the peak value increased a little bit.
Want to ask for advice

**Reply by verseve** [reply]
Hi @Xervou , Could you please start a new Discussion as this question already has been answered? A KsatHorFrac  value of 10,000 is very high, generally we use an upper value of 1000. Additionally, it would be good if you could provide a bit more detail about you wflow model setup (for example catchment, forcing, setup with hydromt_wflow?, etc.).
